### ISSUE TITLE
Certain visualization_settings can cause empty/missing columns on export

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#19889]
**Describe the bug**
Certain visualization_settings can cause empty/missing columns on export.
Very similar to #19465 (#17753)

**To Reproduce**
1. Native query > Sample Dataset > `select 1 a, 2 b, 3 c` - run query, save question
![image](https://user-images.githubusercontent.com/1447303/150858403-d0c96a21-f547-4015-a354-dd8c64ab2c6d.png)
2. Visualization settings > change the order of the column, C A B - run query, save question
![image](https://user-images.githubusercontent.com/1447303/150858321-7c5f5d96-4795-4999-8898-ced236ec1edc.png)
3. Change the query to `select 1 x, 2 y, 3 c` - do **not** run query, save question - this is likely some thing that Anton fixed in 40/41, where it would update the columns, when you ran the query, but that's not important, the empty export is important.
4. After question is saved, do a browser refresh and export (CSV/XLSX), which now only contains `c` column, since we only renamed `a` and `b`. If we had also renamed `c`=>`z`, then export would have been completely empty.
![image](https://user-images.githubusercontent.com/1447303/150859389-4ac23d2f-c06a-4577-a370-bad13af72af7.png)
![image](https://user-images.githubusercontent.com/1447303/150859570-1c227a48-0363-4ed7-870d-90151ceab3f9.png)

Basically the reproduction is a question with the following:
- query = `select 1 x, 2 y, 3 c`
- visualization_settings = `{"table.pivot_column":"A","table.cell_column":"B","table.columns":[{"name":"C","fieldRef":["field","C",{"base-type":"type/Integer"}],"enabled":true},{"name":"A","fieldRef":["field","A",{"base-type":"type/Integer"}],"enabled":true},{"name":"B","fieldRef":["field","B",{"base-type":"type/Integer"}],"enabled":true}]}`

**Information about your Metabase Installation:**
0.41.6

**Additional context**
Thanks to @ds-clearago from #19717

**Comment by noahmoss** [comment]
(oops, hit "close issue" by accident while I was typing up a comment. ignore that)

This is pretty interesting. During export generation on the backend, the column metadata in the query results contain:
```
({:display_name "X", :source :native, :name "X", :base_type :type/Integer}
 {:display_name "Y", :source :native, :name "Y", :base_type :type/Integer}
 {:display_name "C", :source :native, :name "C", :base_type :type/Integer})
 ```
 
 while in the viz settings in the DB, the column names are still `C`, `A` and `B`.
 
Basically, changing the query and saving the question without re-running saves the _old_ viz settings to the DB, with outdated column names. This means that the backend has no way to correlate the columns between the viz settings order and the actual query results, which has new column names.

It seems that modifying the SQL (even just adding and then deleting whitespace), then running the query and re-saving it fixes the column names in the viz settings and thus fixes the export as well, so that is a workaround for anyone who runs into this.

Ultimately I think this is a frontend bug -- when the SQL is edited & re-saved, the viz settings need to be updated regardless of whether the query was run. This might be very difficult to fix, though. Let me think about if there are any alternatives.

**Comment by flamber** [comment]
@noahmoss While I agree with you going forward, how should the frontend be able to handle this for existing questions.
It's also possible to make change in Notebook editor and never visualize, but just click Save question, which will result in similar behavior.

I know how I would fix this in PHP, so I'm sure that it can be done in Clojure. Basically ignore any `"table.columns"` which are not found in the result metadata, since they are just orphaned and can be skipped.

**Comment by noahmoss** [comment]
@flamber isn't that the current behavior? Columns `A` and `B` in `table.columns` are orphaned, so they are skipped. `C` is found in the results metadata, so it is included.

We don't include extra columns in the result metadata that _aren't_ in the viz settings, so `X` and `Y` are not included.

Are you able to reproduce this with the Notebook editor? I don't see it happening if I make a change to the query and save without visualizing.

**Comment by noahmoss** [comment]
I think an alternate way to fix this would be to just have the frontend include the current viz settings in the export request for saved native queries (the same thing we do for unsaved queries of any type).

**Comment by flamber** [comment]
@noahmoss 

Some have many thousands of older questions, where `table.columns` might not match the results. Telling people to re-run and re-save all their questions via the GUI manually is not feasible. If we had a one-click-tool that did this, then yeah, I would totally be with you, but we don't, so we need to fix this.

If there's an orphaned column in `table.columns`, then ignore it - ignore the specific `table.columns`, since it means nothing, but **not** skip it from the result.
We should include all columns from the result. Only exclude the columns that exists in `table.columns` where `enabled: false`

If someone is using the API, then no fixes to the frontend will fix this problem.

**Comment by noahmoss** [comment]
> We should include all columns from the result. Only exclude the columns that exists in table.columns where enabled: false

The problem with this is that we don't know how to order columns from the result if they're not in `table.columns`. We could append them to the beginning or end of each row but it'll always be incorrect for some cases.

I just noticed that when you edit a native query and save it without running it, it resets the column order back to the order specified in the query. So I think we can actually treat this as a special case of [#19796](https://github.com/metabase/metabase/issues/19796): if _any_ viz setting columns can't be correlated with columns from the results, we assume viz settings are out of date or busted, and we fallback to using the original order of the results.
